### PR TITLE
[JUJU-3648] Model table domain

### PR DESCRIPTION
**Pull request by SimonRichardson**
~~Requires https://github.com/juju/juju/pull/15568 to land first. Will rebase after that's landed.~~

---

This patch is broken off from https://github.com/juju/juju/pull/15539. The following implements the
model list via the model manager domain. The purpose of this set
of changes is to ensure that we restrict how databases are created.
This should then limit free-form database creations which could cause
problems at scale. The solution is to add the model to the controller
model list and when requesting to access that model, that the uuid
exists in that table.

The code is simple, push the creation through the modelmanager facade
and call the create and delete at the correct location.

Destroying a model will also remove the model from the controller model
list table. If the model fails to be removed from the db, but is removed
from state, then it will still be cleaned up. The model uuid will be
tainted and can not be used again. This should be fine as model uuid are
randomised.

We should do something about UUIDs in juju there are 6 packages for uuid
imported. We should pick one and use that. I'm currently using utils/uuid
but would rather use an external one, or anything not within the utils
package.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Adding a model should show the new model.

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ make repl-install repl
$ dqlite -s $(sudo cat /var/lib/juju/dqlite/info.yaml | grep "Address: " | cut -d" " -f2) controller
$ SELECT * FROM model_list;
```

Destroying a model should remove the model.

```sh
$ juju destroy-model default --no-prompt
$ make repl-install repl
$ dqlite -s $(sudo cat /var/lib/juju/dqlite/info.yaml | grep "Address: " | cut -d" " -f2) controller
$ SELECT * FROM model_list;
```

### CODE REVIEW

File: database/errors.go [hunk 1]
```diff
@@ -32,3 +33,13 @@ func IsErrConstraintUnique(err error) bool {
...
 }
+
+// IsErrNotFound returns true if the input error was returned by SQLite due
+// to a missing record.
+func IsErrNotFound(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	return errors.Is(err, sql.ErrNoRows)
```
Comment (manadart) [inline database/errors.go:44]: Just verify that we don't need to check for Dqlite error types, as in https://github.com/juju/juju/pull/15547.

File: provider/dummy/environs.go [hunk 2]
```diff
@@ -233,6 +236,7 @@ type environProvider struct {
 	apiPort                    int
 	controllerState            *environState
 	state                      map[string]*environState
+	db                         coredatabase.TrackedDB
```
Comment (manadart) [inline provider/dummy/environs.go:239]: I really hate having to mess about with the dummy environ. It should be strangled out.

File: apiserver/facades/client/modelmanager/modelinfo_test.go [hunk 3]
```diff
@@ -1338,3 +1340,13 @@ func (m *mockMigration) StartTime() time.Time {
 func (m *mockMigration) EndTime() time.Time {
 	return m.end
 }
+
+type mockModelManagerService struct{}
```
Comment (manadart) [inline apiserver/facades/client/modelmanager/modelinfo_test.go:1344]: We might as well generate a mock for this. This is really a stub, as we're not verifying the calls anywhere. Ultimately we'll have to.

File: domain/state.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/juju/juju/core/database"
+)
+
+// DBFactory defines a function that returns a database or an error if a
+// database cannot be returned.
+type DBFactory = func() (database.TrackedDB, error)
+
+// NamespaceDBFactory returns a DBFactory that returns a database from the
+// given getter function.
+func NamespaceDBFactory(getter func(string) (database.TrackedDB, error), namespace string) DBFactory {
```
Comment (manadart) [inline domain/state.go:20]: Not convinced we'll need this. Can we drop it from this patch?

File: domain/modelmanager/state/state.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		stmt := "DELETE FROM model_list WHERE uuid = ?;"
+		result, err := tx.ExecContext(ctx, stmt, uuid)
+		if err != nil {
+			return domain.CoerceError(err)
```
Comment (manadart) [inline domain/modelmanager/state/state.go:66]: I think it's OK for the state layer to emit DB-specific errors and for the service layer to do coercion.

**Review by manadart — CHANGES_REQUESTED** [review]
Couple of suggestions.

We could also have divided this further - just the domain additions could be a patch for example.

File: database/errors.go (hunk 1 above)
Comment (SimonRichardson) [reply database/errors.go:44]: There was nothing from [https://github.com/mattn/go-sqlite3/blob/master/error.go#L37](here) that jumped out.

File: provider/dummy/environs.go (hunk 2 above)
Comment (SimonRichardson) [reply provider/dummy/environs.go:239]: 100% yes. You're not actually testing what you think you're testing at this level.

File: apiserver/facades/client/modelmanager/modelinfo_test.go (hunk 3 above)
Comment (SimonRichardson) [reply apiserver/facades/client/modelmanager/modelinfo_test.go:1344]: Although I am in agreement that this should be a mock, I think for this patch we should just replicate what we have. The whole model manager needs a re-work from a testing perspective and it seems like a big task to just do that in this patch.

File: domain/state.go (hunk 4 above)
Comment (SimonRichardson) [reply domain/state.go:20]: Sure.

File: domain/modelmanager/state/state.go (hunk 5 above)
Comment (SimonRichardson) [reply domain/modelmanager/state/state.go:66]: That actually feels a lot more sane.

File: domain/errors.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package domain
+
+import (
+	"github.com/juju/errors"
```
Comment (manadart) [inline domain/errors.go:7]: Separate stanza.

**Review by manadart — APPROVED** [review]
OK to land this as the first step toward DB name-spacing by model UUID.

We just have to be very disciplined that we roll up the other aspects under the same API methods and poke them into the transactions we've created here.

**Pull request merged**